In [1]:
import pandas as pd
import numpy as np

## Ekstraksi Data bertipe Tanggal - DateTime

In [2]:
df = pd.read_csv('data.csv')
df.head()

,receipt_id,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity
0,9622257,32369294,7/22/2018 21:19,Rice,Rice,supermarket,128000.0,0,1
1,9446359,31885876,7/15/2018 16:17,Rice,Rice,minimarket,102750.0,0,1
2,9470290,31930241,7/15/2018 12:12,Rice,Rice,supermarket,64000.0,0,3
3,9643416,32418582,7/24/2018 8:27,Rice,Rice,minimarket,65000.0,0,1
4,9692093,32561236,7/26/2018 11:28,Rice,Rice,supermarket,124500.0,0,1


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72000 entries, 0 to 71999
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   receipt_id        72000 non-null  int64  
 1   receipts_item_id  72000 non-null  int64  
 2   purchase_time     72000 non-null  object 
 3   category          72000 non-null  object 
 4   sub_category      72000 non-null  object 
 5   format            72000 non-null  object 
 6   unit_price        72000 non-null  float64
 7   discount          72000 non-null  int64  
 8   quantity          72000 non-null  int64  
dtypes: float64(1), int64(4), object(4)
memory usage: 4.9+ MB


## Kolom tanggal, Hanya dapat di ekstraksi jika bertipe data **Datetime**
- Secara default Pandas akan membaca Kolom selain Numerik sebagai Object, Sehingga perlu kita konversi

In [4]:
## Konversi Data

## Alt 1
- Menggunakan parameter **parse_dates = [Nama Kolom]** Ketika awal membaca file

In [5]:
df = pd.read_csv('data.csv', index_col = 0, parse_dates=['purchase_time'])
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity
receipt_id,,,,,,,,
9622257,32369294,2018-07-22 21:19:00,Rice,Rice,supermarket,128000.0,0,1
9446359,31885876,2018-07-15 16:17:00,Rice,Rice,minimarket,102750.0,0,1
9470290,31930241,2018-07-15 12:12:00,Rice,Rice,supermarket,64000.0,0,3
9643416,32418582,2018-07-24 08:27:00,Rice,Rice,minimarket,65000.0,0,1
9692093,32561236,2018-07-26 11:28:00,Rice,Rice,supermarket,124500.0,0,1


In [6]:
df.dtypes

receipts_item_id             int64
purchase_time       datetime64[ns]
category                    object
sub_category                object
format                      object
unit_price                 float64
discount                     int64
quantity                     int64
dtype: object

## Alt 2
- menggu7nakan fungsi .astype()
- astype digunakan untuk mengubah tipe data, tidak terbatas pada date time
- .astype(datetime)

In [7]:
df = pd.read_csv('data.csv', index_col =0)
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity
receipt_id,,,,,,,,
9622257,32369294,7/22/2018 21:19,Rice,Rice,supermarket,128000.0,0,1
9446359,31885876,7/15/2018 16:17,Rice,Rice,minimarket,102750.0,0,1
9470290,31930241,7/15/2018 12:12,Rice,Rice,supermarket,64000.0,0,3
9643416,32418582,7/24/2018 8:27,Rice,Rice,minimarket,65000.0,0,1
9692093,32561236,7/26/2018 11:28,Rice,Rice,supermarket,124500.0,0,1


In [8]:
df.dtypes

receipts_item_id      int64
purchase_time        object
category             object
sub_category         object
format               object
unit_price          float64
discount              int64
quantity              int64
dtype: object

In [9]:
df['purchase_time'].astype('datetime64[ns]')

receipt_id
9622257   2018-07-22 21:19:00
9446359   2018-07-15 16:17:00
9470290   2018-07-15 12:12:00
9643416   2018-07-24 08:27:00
9692093   2018-07-26 11:28:00
                  ...        
5909305   2017-12-27 09:20:00
5736299   2017-12-13 19:52:00
5901144   2017-12-27 08:03:00
5660630   2017-12-07 12:29:00
5808147   2017-12-19 18:59:00
Name: purchase_time, Length: 72000, dtype: datetime64[ns]

In [10]:
df['purchase_time'] = df['purchase_time'].astype('datetime64[ns]')

In [11]:
df.dtypes

receipts_item_id             int64
purchase_time       datetime64[ns]
category                    object
sub_category                object
format                      object
unit_price                 float64
discount                     int64
quantity                     int64
dtype: object

## Alt 3
- Menggunakan fungsi **pd.to_datetime(nama kolom)**

In [12]:
df = pd.read_csv('data.csv', index_col =0)
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity
receipt_id,,,,,,,,
9622257,32369294,7/22/2018 21:19,Rice,Rice,supermarket,128000.0,0,1
9446359,31885876,7/15/2018 16:17,Rice,Rice,minimarket,102750.0,0,1
9470290,31930241,7/15/2018 12:12,Rice,Rice,supermarket,64000.0,0,3
9643416,32418582,7/24/2018 8:27,Rice,Rice,minimarket,65000.0,0,1
9692093,32561236,7/26/2018 11:28,Rice,Rice,supermarket,124500.0,0,1


In [13]:
df.dtypes

receipts_item_id      int64
purchase_time        object
category             object
sub_category         object
format               object
unit_price          float64
discount              int64
quantity              int64
dtype: object

In [14]:
pd.to_datetime(df['purchase_time'])

receipt_id
9622257   2018-07-22 21:19:00
9446359   2018-07-15 16:17:00
9470290   2018-07-15 12:12:00
9643416   2018-07-24 08:27:00
9692093   2018-07-26 11:28:00
                  ...        
5909305   2017-12-27 09:20:00
5736299   2017-12-13 19:52:00
5901144   2017-12-27 08:03:00
5660630   2017-12-07 12:29:00
5808147   2017-12-19 18:59:00
Name: purchase_time, Length: 72000, dtype: datetime64[ns]

In [15]:
df['purchase_time'] = pd.to_datetime(df['purchase_time'])

In [16]:
df.dtypes

receipts_item_id             int64
purchase_time       datetime64[ns]
category                    object
sub_category                object
format                      object
unit_price                 float64
discount                     int64
quantity                     int64
dtype: object

In [17]:
tanggal = pd.Series([
'31-03-2022',
'01-04-2022',
'02-04-2022',
'17-08-2022',
'31-10-2022',
'03-04-2022'])

In [18]:
tanggal

0    31-03-2022
1    01-04-2022
2    02-04-2022
3    17-08-2022
4    31-10-2022
5    03-04-2022
dtype: object

In [19]:
import warnings
warnings.filterwarnings('ignore')

In [20]:
tanggal.astype('datetime64[ns]')

0   2022-03-31
1   2022-04-01
2   2022-04-02
3   2022-08-17
4   2022-10-31
5   2022-04-03
dtype: datetime64[ns]

In [21]:
pd.to_datetime(tanggal)

0   2022-03-31
1   2022-04-01
2   2022-04-02
3   2022-08-17
4   2022-10-31
5   2022-04-03
dtype: datetime64[ns]

In [22]:
tgl = pd.Series([
'01-maret-2005',
'july/17/2017',
'Agustus-20-2021',
'20-september-2003',
'19-August-2016',
'20/May/2004',
'2021-02-15',
'07/25/2000',
'24-12-22',
'januari-01-2023'])
tgl

0        01-maret-2005
1         july/17/2017
2      Agustus-20-2021
3    20-september-2003
4       19-August-2016
5          20/May/2004
6           2021-02-15
7           07/25/2000
8             24-12-22
9      januari-01-2023
dtype: object

In [23]:
#tgl.astype('datetime64[ns]')
## Error

In [24]:
#pd.to_datetime(tgl, yearfirst = True)
#pd.to_datetime(tgl, dayfirst=True)

In [25]:
import dateparser

In [26]:
#!pip install dateparser

Defaulting to user installation because normal site-packages is not writeable


In [27]:
tgl

0        01-maret-2005
1         july/17/2017
2      Agustus-20-2021
3    20-september-2003
4       19-August-2016
5          20/May/2004
6           2021-02-15
7           07/25/2000
8             24-12-22
9      januari-01-2023
dtype: object

In [28]:
tgl.apply(dateparser.parse)

0   2005-03-01
1   2017-07-17
2   2021-08-20
3   2003-09-20
4   2016-08-19
5   2004-05-20
6   2021-02-15
7   2000-07-25
8   2022-12-24
9   2023-01-01
dtype: datetime64[ns]

## Proses Ekstraksi

In [29]:
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity
receipt_id,,,,,,,,
9622257,32369294,2018-07-22 21:19:00,Rice,Rice,supermarket,128000.0,0,1
9446359,31885876,2018-07-15 16:17:00,Rice,Rice,minimarket,102750.0,0,1
9470290,31930241,2018-07-15 12:12:00,Rice,Rice,supermarket,64000.0,0,3
9643416,32418582,2018-07-24 08:27:00,Rice,Rice,minimarket,65000.0,0,1
9692093,32561236,2018-07-26 11:28:00,Rice,Rice,supermarket,124500.0,0,1


In [30]:
df.dtypes

receipts_item_id             int64
purchase_time       datetime64[ns]
category                    object
sub_category                object
format                      object
unit_price                 float64
discount                     int64
quantity                     int64
dtype: object

In [31]:
purchase_dt = df['purchase_time'].dt

In [32]:
### Ambil - Ekstrak Data Bulan berupa Angka
purchase_dt.month

receipt_id
9622257     7
9446359     7
9470290     7
9643416     7
9692093     7
           ..
5909305    12
5736299    12
5901144    12
5660630    12
5808147    12
Name: purchase_time, Length: 72000, dtype: int32

In [33]:
## Memasukkan data hasil ekstraksi ke dalam dataframe

In [34]:
df['month'] = purchase_dt.month

In [35]:
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month
receipt_id,,,,,,,,,
9622257,32369294,2018-07-22 21:19:00,Rice,Rice,supermarket,128000.0,0,1,7
9446359,31885876,2018-07-15 16:17:00,Rice,Rice,minimarket,102750.0,0,1,7
9470290,31930241,2018-07-15 12:12:00,Rice,Rice,supermarket,64000.0,0,3,7
9643416,32418582,2018-07-24 08:27:00,Rice,Rice,minimarket,65000.0,0,1,7
9692093,32561236,2018-07-26 11:28:00,Rice,Rice,supermarket,124500.0,0,1,7


In [36]:
## Ambil data bulan berupa nama
purchase_dt.month_name()

receipt_id
9622257        July
9446359        July
9470290        July
9643416        July
9692093        July
             ...   
5909305    December
5736299    December
5901144    December
5660630    December
5808147    December
Name: purchase_time, Length: 72000, dtype: object

In [37]:
## Memasukkan hasil ekstraksi ke dataframe
df['bulan'] = purchase_dt.month_name()

In [38]:
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month,bulan
receipt_id,,,,,,,,,,
9622257,32369294,2018-07-22 21:19:00,Rice,Rice,supermarket,128000.0,0,1,7,July
9446359,31885876,2018-07-15 16:17:00,Rice,Rice,minimarket,102750.0,0,1,7,July
9470290,31930241,2018-07-15 12:12:00,Rice,Rice,supermarket,64000.0,0,3,7,July
9643416,32418582,2018-07-24 08:27:00,Rice,Rice,minimarket,65000.0,0,1,7,July
9692093,32561236,2018-07-26 11:28:00,Rice,Rice,supermarket,124500.0,0,1,7,July


In [39]:
## ambil data Tahun
purchase_dt.year

receipt_id
9622257    2018
9446359    2018
9470290    2018
9643416    2018
9692093    2018
           ... 
5909305    2017
5736299    2017
5901144    2017
5660630    2017
5808147    2017
Name: purchase_time, Length: 72000, dtype: int32

In [40]:
df['year'] = purchase_dt.year

In [41]:
## Ambil data Tanggal
purchase_dt.day

receipt_id
9622257    22
9446359    15
9470290    15
9643416    24
9692093    26
           ..
5909305    27
5736299    13
5901144    27
5660630     7
5808147    19
Name: purchase_time, Length: 72000, dtype: int32

In [42]:
## Ambil Hari - berupa angka
purchase_dt.dayofweek
## Berurutan senin - minggu, senin = 0, minggu = 6

receipt_id
9622257    6
9446359    6
9470290    6
9643416    1
9692093    3
          ..
5909305    2
5736299    2
5901144    2
5660630    3
5808147    1
Name: purchase_time, Length: 72000, dtype: int32

In [43]:
## Ambil Hari - berupa Nama
purchase_dt.day_name()

receipt_id
9622257       Sunday
9446359       Sunday
9470290       Sunday
9643416      Tuesday
9692093     Thursday
             ...    
5909305    Wednesday
5736299    Wednesday
5901144    Wednesday
5660630     Thursday
5808147      Tuesday
Name: purchase_time, Length: 72000, dtype: object

In [44]:
## memasukkan data ke Dataframe
df['day'] = purchase_dt.day_name()

In [45]:
df['hari'] = purchase_dt.dayofweek

In [46]:
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month,bulan,year,day,hari
receipt_id,,,,,,,,,,,,,
9622257,32369294,2018-07-22 21:19:00,Rice,Rice,supermarket,128000.0,0,1,7,July,2018,Sunday,6
9446359,31885876,2018-07-15 16:17:00,Rice,Rice,minimarket,102750.0,0,1,7,July,2018,Sunday,6
9470290,31930241,2018-07-15 12:12:00,Rice,Rice,supermarket,64000.0,0,3,7,July,2018,Sunday,6
9643416,32418582,2018-07-24 08:27:00,Rice,Rice,minimarket,65000.0,0,1,7,July,2018,Tuesday,1
9692093,32561236,2018-07-26 11:28:00,Rice,Rice,supermarket,124500.0,0,1,7,July,2018,Thursday,3


In [47]:
def WeekEnd(x):
    if x > 4:
        return "Yes"
    else:
        return "No"

In [48]:
df['hari'].apply(WeekEnd)

receipt_id
9622257    Yes
9446359    Yes
9470290    Yes
9643416     No
9692093     No
          ... 
5909305     No
5736299     No
5901144     No
5660630     No
5808147     No
Name: hari, Length: 72000, dtype: object

In [49]:
df['WeekEnd'] = df['hari'].apply(WeekEnd)

In [50]:
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month,bulan,year,day,hari,WeekEnd
receipt_id,,,,,,,,,,,,,,
9622257,32369294,2018-07-22 21:19:00,Rice,Rice,supermarket,128000.0,0,1,7,July,2018,Sunday,6,Yes
9446359,31885876,2018-07-15 16:17:00,Rice,Rice,minimarket,102750.0,0,1,7,July,2018,Sunday,6,Yes
9470290,31930241,2018-07-15 12:12:00,Rice,Rice,supermarket,64000.0,0,3,7,July,2018,Sunday,6,Yes
9643416,32418582,2018-07-24 08:27:00,Rice,Rice,minimarket,65000.0,0,1,7,July,2018,Tuesday,1,No
9692093,32561236,2018-07-26 11:28:00,Rice,Rice,supermarket,124500.0,0,1,7,July,2018,Thursday,3,No


In [51]:
## Ambil Data Jam
purchase_dt.hour

receipt_id
9622257    21
9446359    16
9470290    12
9643416     8
9692093    11
           ..
5909305     9
5736299    19
5901144     8
5660630    12
5808147    18
Name: purchase_time, Length: 72000, dtype: int32

In [52]:
## ambil hanya data tanggal (date)
purchase_dt.to_period('D')

receipt_id
9622257    2018-07-22
9446359    2018-07-15
9470290    2018-07-15
9643416    2018-07-24
9692093    2018-07-26
              ...    
5909305    2017-12-27
5736299    2017-12-13
5901144    2017-12-27
5660630    2017-12-07
5808147    2017-12-19
Name: purchase_time, Length: 72000, dtype: period[D]

In [53]:
## Ambil data Bulan - Tahun
purchase_dt.to_period('M')

receipt_id
9622257    2018-07
9446359    2018-07
9470290    2018-07
9643416    2018-07
9692093    2018-07
            ...   
5909305    2017-12
5736299    2017-12
5901144    2017-12
5660630    2017-12
5808147    2017-12
Name: purchase_time, Length: 72000, dtype: period[M]

In [54]:
## Ambil data Weekly - Periode Mingguan (senin - minggu)

In [55]:
purchase_dt.to_period('W')

receipt_id
9622257    2018-07-16/2018-07-22
9446359    2018-07-09/2018-07-15
9470290    2018-07-09/2018-07-15
9643416    2018-07-23/2018-07-29
9692093    2018-07-23/2018-07-29
                   ...          
5909305    2017-12-25/2017-12-31
5736299    2017-12-11/2017-12-17
5901144    2017-12-25/2017-12-31
5660630    2017-12-04/2017-12-10
5808147    2017-12-18/2017-12-24
Name: purchase_time, Length: 72000, dtype: period[W-SUN]

In [56]:
## Ambil data Kuartal
purchase_dt.to_period('Q')

receipt_id
9622257    2018Q3
9446359    2018Q3
9470290    2018Q3
9643416    2018Q3
9692093    2018Q3
            ...  
5909305    2017Q4
5736299    2017Q4
5901144    2017Q4
5660630    2017Q4
5808147    2017Q4
Name: purchase_time, Length: 72000, dtype: period[Q-DEC]

In [57]:
df['quarter'] = purchase_dt.to_period('Q')

In [58]:
df['yearMonth'] = purchase_dt.to_period('M')

In [59]:
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month,bulan,year,day,hari,WeekEnd,quarter,yearMonth
receipt_id,,,,,,,,,,,,,,,,
9622257,32369294,2018-07-22 21:19:00,Rice,Rice,supermarket,128000.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07
9446359,31885876,2018-07-15 16:17:00,Rice,Rice,minimarket,102750.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07
9470290,31930241,2018-07-15 12:12:00,Rice,Rice,supermarket,64000.0,0,3,7,July,2018,Sunday,6,Yes,2018Q3,2018-07
9643416,32418582,2018-07-24 08:27:00,Rice,Rice,minimarket,65000.0,0,1,7,July,2018,Tuesday,1,No,2018Q3,2018-07
9692093,32561236,2018-07-26 11:28:00,Rice,Rice,supermarket,124500.0,0,1,7,July,2018,Thursday,3,No,2018Q3,2018-07


In [60]:
df['tgl'] = purchase_dt.day

In [61]:
df['hour'] = purchase_dt.hour

In [62]:
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month,bulan,year,day,hari,WeekEnd,quarter,yearMonth,tgl,hour
receipt_id,,,,,,,,,,,,,,,,,,
9622257,32369294,2018-07-22 21:19:00,Rice,Rice,supermarket,128000.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,22,21
9446359,31885876,2018-07-15 16:17:00,Rice,Rice,minimarket,102750.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,16
9470290,31930241,2018-07-15 12:12:00,Rice,Rice,supermarket,64000.0,0,3,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,12
9643416,32418582,2018-07-24 08:27:00,Rice,Rice,minimarket,65000.0,0,1,7,July,2018,Tuesday,1,No,2018Q3,2018-07,24,8
9692093,32561236,2018-07-26 11:28:00,Rice,Rice,supermarket,124500.0,0,1,7,July,2018,Thursday,3,No,2018Q3,2018-07,26,11


In [63]:
df['total_payment'] = (df['unit_price'] * df['quantity']) - df['discount']

In [64]:
purchase_dt.strftime('%B / %H:%M')
## Custom format sesuai kebutuhan
## Format mengacu pada format Date time pada python ==> strftime.org

receipt_id
9622257        July / 21:19
9446359        July / 16:17
9470290        July / 12:12
9643416        July / 08:27
9692093        July / 11:28
                 ...       
5909305    December / 09:20
5736299    December / 19:52
5901144    December / 08:03
5660630    December / 12:29
5808147    December / 18:59
Name: purchase_time, Length: 72000, dtype: object

In [65]:
df['custom'] = purchase_dt.strftime('%B / %H:%M')

In [66]:
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month,bulan,year,day,hari,WeekEnd,quarter,yearMonth,tgl,hour,total_payment,custom
receipt_id,,,,,,,,,,,,,,,,,,,,
9622257,32369294,2018-07-22 21:19:00,Rice,Rice,supermarket,128000.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,22,21,128000.0,July / 21:19
9446359,31885876,2018-07-15 16:17:00,Rice,Rice,minimarket,102750.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,16,102750.0,July / 16:17
9470290,31930241,2018-07-15 12:12:00,Rice,Rice,supermarket,64000.0,0,3,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,12,192000.0,July / 12:12
9643416,32418582,2018-07-24 08:27:00,Rice,Rice,minimarket,65000.0,0,1,7,July,2018,Tuesday,1,No,2018Q3,2018-07,24,8,65000.0,July / 08:27
9692093,32561236,2018-07-26 11:28:00,Rice,Rice,supermarket,124500.0,0,1,7,July,2018,Thursday,3,No,2018Q3,2018-07,26,11,124500.0,July / 11:28


In [67]:
## Analisis Data

- Crosstab
- Group by
- Pivot Table

In [68]:
## Frequency Table

In [69]:
## Univariate Analysis -= Analisis per Kolom

In [70]:
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month,bulan,year,day,hari,WeekEnd,quarter,yearMonth,tgl,hour,total_payment,custom
receipt_id,,,,,,,,,,,,,,,,,,,,
9622257,32369294,2018-07-22 21:19:00,Rice,Rice,supermarket,128000.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,22,21,128000.0,July / 21:19
9446359,31885876,2018-07-15 16:17:00,Rice,Rice,minimarket,102750.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,16,102750.0,July / 16:17
9470290,31930241,2018-07-15 12:12:00,Rice,Rice,supermarket,64000.0,0,3,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,12,192000.0,July / 12:12
9643416,32418582,2018-07-24 08:27:00,Rice,Rice,minimarket,65000.0,0,1,7,July,2018,Tuesday,1,No,2018Q3,2018-07,24,8,65000.0,July / 08:27
9692093,32561236,2018-07-26 11:28:00,Rice,Rice,supermarket,124500.0,0,1,7,July,2018,Thursday,3,No,2018Q3,2018-07,26,11,124500.0,July / 11:28


## 1. Business Question : Sub Category apa yg memiliki Penjualan (items) Tertinggi?

In [71]:
df['sub_category'].value_counts()

sub_category
Detergent    36000
Sugar        24000
Rice         12000
Name: count, dtype: int64

In [72]:
pd.crosstab(df['sub_category'], columns='Jumlah')

col_0,Jumlah
sub_category,
Detergent,36000
Rice,12000
Sugar,24000


- **Insight** - Informasi : Penjualan tertinggi (items) dari Sub category detergent sebanyak 36.000 transaksi

## 2. Business Question : Sub Category apa yg memiliki penjualan (items) paling sedikit?

In [73]:
df['sub_category'].value_counts()

sub_category
Detergent    36000
Sugar        24000
Rice         12000
Name: count, dtype: int64

In [74]:
pd.crosstab(df['sub_category'], columns='Jumlah').sort_values('Jumlah')

col_0,Jumlah
sub_category,
Rice,12000
Sugar,24000
Detergent,36000


- **Insight** - Informasi : Penjualan paling sedikit (items) dari sub category Rice sebanyak 12.000 transaksi

## 3. Business Question : Lokasi (format) mana yg memiliki penjualan paling sedikit?

In [75]:
df['format'].value_counts()

format
minimarket     46803
supermarket    19826
hypermarket     5371
Name: count, dtype: int64

- Insight - Informasi :
    - Penjualan paling sedikit (items) dari lokasi/format Hypermarket sebanyak 5371
    - Penjualan paling tinggi ada di Minimarket sebanyak 46.803, karena lokasi minimarket dekat dg perumahan dan jumlahnya cukup banyak. rata-rata jarakl 100m dari perumahan, jumlah minimal ada 2 minimarket
    
source : https://bisnis.co.id/rata-rata-jarak-minimarket

In [76]:
pd.crosstab(df['format'], columns='Jumlah').sort_values('Jumlah')

col_0,Jumlah
format,
hypermarket,5371
supermarket,19826
minimarket,46803


## Insight Data - Informasi :
- Transaksi paling tinggi ada di minimarket, sedangkan paling rendah ada di hypermarket
- Jika akan merilis produk baru, sebaikonya dilakukan di minimarket
- Jika akan melakukan promo produk sebaiknya di minimarket
- Penjualan di minimarket tinggi karena lokasi minimarket dekat dg perumahan dan jumlahnya cukup banyak. rata-rata jarakl 100m dari perumahan, jumlah minimal ada 2 minimarket ==> Hipotesa/asumsi
- Jika ada Hipotesa - dugaan - asumsi harus berdasarkan data

Source : .....

In [77]:
pd.crosstab(index=df['category'], columns='Jumlah')

col_0,Jumlah
category,
Fabric Care,36000
Rice,12000
Sugar/Flavored Syrup,24000


In [78]:
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month,bulan,year,day,hari,WeekEnd,quarter,yearMonth,tgl,hour,total_payment,custom
receipt_id,,,,,,,,,,,,,,,,,,,,
9622257,32369294,2018-07-22 21:19:00,Rice,Rice,supermarket,128000.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,22,21,128000.0,July / 21:19
9446359,31885876,2018-07-15 16:17:00,Rice,Rice,minimarket,102750.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,16,102750.0,July / 16:17
9470290,31930241,2018-07-15 12:12:00,Rice,Rice,supermarket,64000.0,0,3,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,12,192000.0,July / 12:12
9643416,32418582,2018-07-24 08:27:00,Rice,Rice,minimarket,65000.0,0,1,7,July,2018,Tuesday,1,No,2018Q3,2018-07,24,8,65000.0,July / 08:27
9692093,32561236,2018-07-26 11:28:00,Rice,Rice,supermarket,124500.0,0,1,7,July,2018,Thursday,3,No,2018Q3,2018-07,26,11,124500.0,July / 11:28


In [79]:
pd.crosstab(index=df['unit_price'], columns='Jumlah')

col_0,Jumlah
unit_price,
2505.0000,1
2510.0000,2
2515.0000,2
2524.6667,1
2525.0000,2
...,...
218000.0000,1
218592.0000,1
218900.0000,1


## Analisa Univariate menggunakan Crosstab
- Kolom yg dianalisa menggunakan crosstab (secara univariate) adalah Kolom Kategorikal (datanya kategorik)
- Kolom yg berisi data numerik continous bisa dilakukan tetapi tidak direkomendasikan, karena akan sulit mengambil insight data/informasi
- Fungsi Aggregate yg digunakan di Crosstab secara default adalah fungsi **Count**
- Output yg dihasilkan (ketika analisa univariate) sama dengan ketika menggunakan fungsi **.value_counts()**
- Perbedaan terletak pada tipe data yg dihasilkan. Crosstab menghasilkan Dataframe, value_counts menghasilkan Series


## Analisa BiVariate - Multivariate : Analisa 2 Koolom atau Lebih

In [80]:
pd.crosstab(index=df['sub_category'], columns=df['format'])

format,hypermarket,minimarket,supermarket
sub_category,,,
Detergent,2611,24345,9044
Rice,999,7088,3913
Sugar,1761,15370,6869


In [81]:
pd.crosstab(index=df['format'], columns=df['sub_category'])

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,2611,999,1761
minimarket,24345,7088,15370
supermarket,9044,3913,6869


In [82]:
pd.crosstab(index=df['sub_category'], columns=df['format'])[['supermarket']]

format,supermarket
sub_category,
Detergent,9044
Rice,3913
Sugar,6869


In [83]:
pd.crosstab(index=df['sub_category'], columns=df['format'])[['hypermarket']]

format,hypermarket
sub_category,
Detergent,2611
Rice,999
Sugar,1761


In [86]:
pd.crosstab(index=df['format'], columns=df['sub_category'])[['Rice']]

sub_category,Rice
format,
hypermarket,999
minimarket,7088
supermarket,3913


In [87]:
pd.crosstab(index=df['format'], columns=df['day'])

day,Friday,Monday,Saturday,Sunday,Thursday,Tuesday,Wednesday
format,,,,,,,
hypermarket,802,573,1016,1235,571,600,574
minimarket,7004,6044,7411,7697,6133,6269,6245
supermarket,2972,2433,3401,3641,2434,2558,2387


In [88]:
pd.crosstab(index=df['sub_category'], columns=df['day'])

day,Friday,Monday,Saturday,Sunday,Thursday,Tuesday,Wednesday
sub_category,,,,,,,
Detergent,5748,4337,6147,6381,4320,4522,4545
Rice,1646,1551,1915,2234,1552,1621,1481
Sugar,3384,3162,3766,3958,3266,3284,3180


In [89]:
pd.crosstab(index=df['day'], columns=df['sub_category'])

sub_category,Detergent,Rice,Sugar
day,,,
Friday,5748,1646,3384
Monday,4337,1551,3162
Saturday,6147,1915,3766
Sunday,6381,2234,3958
Thursday,4320,1552,3266
Tuesday,4522,1621,3284
Wednesday,4545,1481,3180


In [90]:
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month,bulan,year,day,hari,WeekEnd,quarter,yearMonth,tgl,hour,total_payment,custom
receipt_id,,,,,,,,,,,,,,,,,,,,
9622257,32369294,2018-07-22 21:19:00,Rice,Rice,supermarket,128000.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,22,21,128000.0,July / 21:19
9446359,31885876,2018-07-15 16:17:00,Rice,Rice,minimarket,102750.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,16,102750.0,July / 16:17
9470290,31930241,2018-07-15 12:12:00,Rice,Rice,supermarket,64000.0,0,3,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,12,192000.0,July / 12:12
9643416,32418582,2018-07-24 08:27:00,Rice,Rice,minimarket,65000.0,0,1,7,July,2018,Tuesday,1,No,2018Q3,2018-07,24,8,65000.0,July / 08:27
9692093,32561236,2018-07-26 11:28:00,Rice,Rice,supermarket,124500.0,0,1,7,July,2018,Thursday,3,No,2018Q3,2018-07,26,11,124500.0,July / 11:28


In [91]:
pd.crosstab(index=df['bulan'], columns=df['format'])

format,hypermarket,minimarket,supermarket
bulan,,,
April,376,3909,1715
August,455,3994,1551
December,427,3939,1634
February,513,3745,1742
January,449,3960,1591
July,356,4293,1351
June,458,3949,1593
March,521,3540,1939
May,487,3704,1809


In [95]:
pd.crosstab(index=df['bulan'], columns=df['format'])[['minimarket']].sort_values('minimarket', ascending=False)

format,minimarket
bulan,
July,4293
August,3994
January,3960
September,3960
June,3949
December,3939
November,3912
April,3909
October,3898


In [99]:
pd.crosstab(index=df['bulan'], columns=df['sub_category'])

sub_category,Detergent,Rice,Sugar
bulan,,,
April,3000,1000,2000
August,3000,1000,2000
December,3000,1000,2000
February,3000,1000,2000
January,3000,1000,2000
July,3000,1000,2000
June,3000,1000,2000
March,3000,1000,2000
May,3000,1000,2000


In [100]:
pd.crosstab(index=df['bulan'], columns=df['sub_category'])[['Sugar']]

sub_category,Sugar
bulan,
April,2000
August,2000
December,2000
February,2000
January,2000
July,2000
June,2000
March,2000
May,2000


In [101]:
pd.crosstab(index=df['bulan'], columns=[df['sub_category'], df['format']])

sub_category   Detergent                               Rice             \
format       hypermarket minimarket supermarket hypermarket minimarket   
bulan                                                                    
April                176       2031         793          79        578   
August               253       2025         722          82        611   
December             206       2111         683          68        583   
February             212       1989         799         117        549   
January              229       2094         677          68        597   
July                 169       2289         542          65        626   
June                 230       1996         774          81        622   
March                249       1853         898          97        559   
May                  236       1884         880          82        585   
November             218       2069         713          76        581   
October              213       2009         778          81        600   
September            220       1995         785         103        597   

sub_category                   Sugar                         
format       supermarket hypermarket minimarket supermarket  
bulan                                                        
April                343         121       1300         579  
August               307         120       1358         522  
December             349         153       1245         602  
February             334         184       1207         609  
January              335         152       1269         579  
July                 309         122       1378         500  
June                 297         147       1331         522  
March                344         175       1128         697  
May                  333         169       1235         596  
November             343         148       1262         590  
October              319         152       1289         559  
September            300         118       1368         514

In [102]:
pd.crosstab(index=df['bulan'], columns=[df['sub_category'], df['format']]).T

bulan                     April  August  December  February  January  July  \
sub_category format                                                          
Detergent    hypermarket    176     253       206       212      229   169   
             minimarket    2031    2025      2111      1989     2094  2289   
             supermarket    793     722       683       799      677   542   
Rice         hypermarket     79      82        68       117       68    65   
             minimarket     578     611       583       549      597   626   
             supermarket    343     307       349       334      335   309   
Sugar        hypermarket    121     120       153       184      152   122   
             minimarket    1300    1358      1245      1207     1269  1378   
             supermarket    579     522       602       609      579   500   

bulan                     June  March   May  November  October  September  
sub_category format                                                        
Detergent    hypermarket   230    249   236       218      213        220  
             minimarket   1996   1853  1884      2069     2009       1995  
             supermarket   774    898   880       713      778        785  
Rice         hypermarket    81     97    82        76       81        103  
             minimarket    622    559   585       581      600        597  
             supermarket   297    344   333       343      319        300  
Sugar        hypermarket   147    175   169       148      152        118  
             minimarket   1331   1128  1235      1262     1289       1368  
             supermarket   522    697   596       590      559        514

In [103]:
pd.crosstab(index=df['bulan'], columns=[df['format'], df['sub_category']])

format       hypermarket            minimarket            supermarket       \
sub_category   Detergent Rice Sugar  Detergent Rice Sugar   Detergent Rice   
bulan                                                                        
April                176   79   121       2031  578  1300         793  343   
August               253   82   120       2025  611  1358         722  307   
December             206   68   153       2111  583  1245         683  349   
February             212  117   184       1989  549  1207         799  334   
January              229   68   152       2094  597  1269         677  335   
July                 169   65   122       2289  626  1378         542  309   
June                 230   81   147       1996  622  1331         774  297   
March                249   97   175       1853  559  1128         898  344   
May                  236   82   169       1884  585  1235         880  333   
November             218   76   148       2069  581  1262         713  343   
October              213   81   152       2009  600  1289         778  319   
September            220  103   118       1995  597  1368         785  300   

format              
sub_category Sugar  
bulan               
April          579  
August         522  
December       602  
February       609  
January        579  
July           500  
June           522  
March          697  
May            596  
November       590  
October        559  
September      514

In [104]:
pd.crosstab(columns=df['bulan'], index=[df['format'], df['sub_category']])

bulan                     April  August  December  February  January  July  \
format      sub_category                                                     
hypermarket Detergent       176     253       206       212      229   169   
            Rice             79      82        68       117       68    65   
            Sugar           121     120       153       184      152   122   
minimarket  Detergent      2031    2025      2111      1989     2094  2289   
            Rice            578     611       583       549      597   626   
            Sugar          1300    1358      1245      1207     1269  1378   
supermarket Detergent       793     722       683       799      677   542   
            Rice            343     307       349       334      335   309   
            Sugar           579     522       602       609      579   500   

bulan                     June  March   May  November  October  September  
format      sub_category                                                   
hypermarket Detergent      230    249   236       218      213        220  
            Rice            81     97    82        76       81        103  
            Sugar          147    175   169       148      152        118  
minimarket  Detergent     1996   1853  1884      2069     2009       1995  
            Rice           622    559   585       581      600        597  
            Sugar         1331   1128  1235      1262     1289       1368  
supermarket Detergent      774    898   880       713      778        785  
            Rice           297    344   333       343      319        300  
            Sugar          522    697   596       590      559        514

In [105]:
pd.crosstab(index=df['day'], columns=[df['sub_category'], df['format']])

sub_category   Detergent                               Rice             \
format       hypermarket minimarket supermarket hypermarket minimarket   
day                                                                      
Friday               398       3884        1466         154        973   
Monday               264       3003        1070         111        958   
Saturday             538       4000        1609         175       1100   
Sunday               598       4097        1686         263       1210   
Thursday             266       3027        1027         100        902   
Tuesday              291       3108        1123          94       1011   
Wednesday            256       3226        1063         102        934   

sub_category                   Sugar                         
format       supermarket hypermarket minimarket supermarket  
day                                                          
Friday               519         250       2147         987  
Monday               482         198       2083         881  
Saturday             640         303       2311        1152  
Sunday               761         374       2390        1194  
Thursday             550         205       2204         857  
Tuesday              516         215       2150         919  
Wednesday            445         216       2085         879

In [106]:
df.columns

Index(['receipts_item_id', 'purchase_time', 'category', 'sub_category',
       'format', 'unit_price', 'discount', 'quantity', 'month', 'bulan',
       'year', 'day', 'hari', 'WeekEnd', 'quarter', 'yearMonth', 'tgl', 'hour',
       'total_payment', 'custom'],
      dtype='object')

In [107]:
pd.crosstab(index=[df['year'], df['quarter'], df['day']], 
            columns=[df['format'], df['category'], df['sub_category']])

format                 hypermarket                            minimarket       \
category               Fabric Care Rice Sugar/Flavored Syrup Fabric Care Rice   
sub_category             Detergent Rice                Sugar   Detergent Rice   
year quarter day                                                                
2017 2017Q4  Friday            109   32                   75         955  245   
             Monday             59   20                   50         733  233   
             Saturday          128   42                   76        1041  280   
             Sunday            152   71                  100        1098  355   
             Thursday           57   19                   49         754  225   
             Tuesday            69   20                   54         765  221   
             Wednesday          63   21                   49         843  205   
2018 2018Q1  Friday            102   57                   62        1016  231   
             Monday             67   28                   55         704  233   
             Saturday          170   52                   96         942  272   
             Sunday            158   59                  122         932  235   
             Thursday           63   33                   58         783  226   
             Tuesday            67   29                   55         772  269   
             Wednesday          63   24                   63         787  239   
     2018Q2  Friday             76   30                   66         942  239   
             Monday             64   27                   53         717  221   
             Saturday          128   47                   74         993  257   
             Sunday            140   69                   81         970  294   
             Thursday           82   20                   51         760  246   
             Tuesday            83   22                   55         789  285   
             Wednesday          69   27                   57         740  243   
     2018Q3  Friday            111   35                   47         971  258   
             Monday             74   36                   40         849  271   
             Saturday          112   34                   57        1024  291   
             Sunday            148   64                   71        1097  326   
             Thursday           64   28                   47         730  205   
             Tuesday            72   23                   51         782  236   
             Wednesday          61   30                   47         856  247   

format                                      supermarket       \
category               Sugar/Flavored Syrup Fabric Care Rice   
sub_category                          Sugar   Detergent Rice   
year quarter day                                               
2017 2017Q4  Friday                     530         336  133   
             Monday                     510         266  114   
             Saturday                   531         362  187   
             Sunday                     641         436  211   
             Thursday                   517         275  127   
             Tuesday                    545         267  131   
             Wednesday                  522         232  108   
2018 2018Q1  Friday                     501         390  141   
             Monday                     510         244  128   
             Saturday                   575         458  151   
             Sunday                     544         446  194   
             Thursday                   529         276  159   
             Tuesday                    489         286  134   
             Wednesday                  456         274  106   
     2018Q2  Friday                     528         382  130   
             Monday                     525         320  111   
             Saturday                   602         420  155   
             Sunday                     565         451  176   

In [108]:
pd.crosstab(index=[df['quarter'], df['format'], df['sub_category']], columns = 'Total Transaksi')

col_0                             Total Transaksi
quarter format      sub_category                 
2017Q4  hypermarket Detergent                 637
                    Rice                      225
                    Sugar                     453
        minimarket  Detergent                6189
                    Rice                     1764
                    Sugar                    3796
        supermarket Detergent                2174
                    Rice                     1011
                    Sugar                    1751
2018Q1  hypermarket Detergent                 690
                    Rice                      282
                    Sugar                     511
        minimarket  Detergent                5936
                    Rice                     1705
                    Sugar                    3604
        supermarket Detergent                2374
                    Rice                     1013
                    Sugar                    1885
2018Q2  hypermarket Detergent                 642
                    Rice                      242
                    Sugar                     437
        minimarket  Detergent                5911
                    Rice                     1785
                    Sugar                    3866
        supermarket Detergent                2447
                    Rice                      973
                    Sugar                    1697
2018Q3  hypermarket Detergent                 642
                    Rice                      250
                    Sugar                     360
        minimarket  Detergent                6309
                    Rice                     1834
                    Sugar                    4104
        supermarket Detergent                2049
                    Rice                      916
                    Sugar                    1536

In [109]:
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month,bulan,year,day,hari,WeekEnd,quarter,yearMonth,tgl,hour,total_payment,custom
receipt_id,,,,,,,,,,,,,,,,,,,,
9622257,32369294,2018-07-22 21:19:00,Rice,Rice,supermarket,128000.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,22,21,128000.0,July / 21:19
9446359,31885876,2018-07-15 16:17:00,Rice,Rice,minimarket,102750.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,16,102750.0,July / 16:17
9470290,31930241,2018-07-15 12:12:00,Rice,Rice,supermarket,64000.0,0,3,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,12,192000.0,July / 12:12
9643416,32418582,2018-07-24 08:27:00,Rice,Rice,minimarket,65000.0,0,1,7,July,2018,Tuesday,1,No,2018Q3,2018-07,24,8,65000.0,July / 08:27
9692093,32561236,2018-07-26 11:28:00,Rice,Rice,supermarket,124500.0,0,1,7,July,2018,Thursday,3,No,2018Q3,2018-07,26,11,124500.0,July / 11:28


In [114]:
df[(df['quarter'] == '2018Q3') & (df['format'] == 'hypermarket')  & (df['sub_category'] == 'Detergent')]

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month,bulan,year,day,hari,WeekEnd,quarter,yearMonth,tgl,hour,total_payment,custom
receipt_id,,,,,,,,,,,,,,,,,,,,
9820180,32977860,2018-07-31 16:38:00,Fabric Care,Detergent,hypermarket,13900.0,0,2,7,July,2018,Tuesday,1,No,2018Q3,2018-07,31,16,27800.0,July / 16:38
9755574,32812018,2018-07-27 12:41:00,Fabric Care,Detergent,hypermarket,25500.0,0,1,7,July,2018,Friday,4,No,2018Q3,2018-07,27,12,25500.0,July / 12:41
9184659,31065015,2018-07-02 16:02:00,Fabric Care,Detergent,hypermarket,20400.0,0,2,7,July,2018,Monday,0,No,2018Q3,2018-07,2,16,40800.0,July / 16:02
9776658,32866909,2018-07-29 19:58:00,Fabric Care,Detergent,hypermarket,7990.0,3110,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,29,19,4880.0,July / 19:58
9751656,32735057,2018-07-28 21:04:00,Fabric Care,Detergent,hypermarket,27990.0,0,2,7,July,2018,Saturday,5,Yes,2018Q3,2018-07,28,21,55980.0,July / 21:04
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10977243,36660176,2018-09-14 17:17:00,Fabric Care,Detergent,hypermarket,11500.0,0,1,9,September,2018,Friday,4,No,2018Q3,2018-09,14,17,11500.0,September / 17:17
11034139,36777090,2018-09-16 21:22:00,Fabric Care,Detergent,hypermarket,22300.0,0,1,9,September,2018,Sunday,6,Yes,2018Q3,2018-09,16,21,22300.0,September / 21:22
11066494,36903749,2018-09-03 17:07:00,Fabric Care,Detergent,hypermarket,13900.0,0,1,9,September,2018,Monday,0,No,2018Q3,2018-09,3,17,13900.0,September / 17:07


In [115]:
pd.crosstab(index=df['format'], columns=df['sub_category'])
## 100%

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,2611,999,1761
minimarket,24345,7088,15370
supermarket,9044,3913,6869


In [116]:
pd.crosstab(index=df['format'], columns=df['sub_category'], normalize=True)

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,0.036264,0.013875,0.024458
minimarket,0.338125,0.098444,0.213472
supermarket,0.125611,0.054347,0.095403


In [117]:
pd.crosstab(index=df['format'], columns=df['sub_category'], normalize=True) * 100
## Normalize = True ==> Jika semua value ditotal, jumlahnya => 1 atau 100%

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,3.626389,1.387500,2.445833
minimarket,33.812500,9.844444,21.347222
supermarket,12.561111,5.434722,9.540278


In [118]:
(pd.crosstab(index=df['format'], columns=df['sub_category'], normalize=True) * 100).round(2)
## fungsi .round(2) => pembulatan 2 angka di belakang koma

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,3.63,1.39,2.45
minimarket,33.81,9.84,21.35
supermarket,12.56,5.43,9.54


In [119]:
3.63 + 1.39 + 2.45 + 33.81 + 9.84 + 21.35 + 12.56 + 5.43 + 9.54

100.0

In [121]:
(pd.crosstab(index=df['format'], columns=df['sub_category'], normalize='all') * 100).round(2)
## Menampilkan hasil yg sama dg normalize = True

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,3.63,1.39,2.45
minimarket,33.81,9.84,21.35
supermarket,12.56,5.43,9.54


In [125]:
pd.crosstab(index=df['format'], columns=df['sub_category'], normalize='columns')
## 100%
## Normalize = 'columns' ==> Values di columns yg sama jika dijumlah = 1 => 100%

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,0.072528,0.083250,0.073375
minimarket,0.676250,0.590667,0.640417
supermarket,0.251222,0.326083,0.286208


In [123]:
pd.crosstab(index=df['format'], columns=df['sub_category'], normalize='columns') * 100

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,7.252778,8.325000,7.337500
minimarket,67.625000,59.066667,64.041667
supermarket,25.122222,32.608333,28.620833


In [124]:
(pd.crosstab(index=df['format'], columns=df['sub_category'], normalize='columns') * 100).round(2)

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,7.25,8.33,7.34
minimarket,67.62,59.07,64.04
supermarket,25.12,32.61,28.62


In [126]:
## Sugar
7.34 + 64.04 + 28.62

100.00000000000001

In [127]:
(pd.crosstab(index=df['format'], columns=df['sub_category'], normalize='columns') * 100).round(2)[['Detergent']]

sub_category,Detergent
format,
hypermarket,7.25
minimarket,67.62
supermarket,25.12


In [128]:
df[df['sub_category'] == 'Detergent']

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month,bulan,year,day,hari,WeekEnd,quarter,yearMonth,tgl,hour,total_payment,custom
receipt_id,,,,,,,,,,,,,,,,,,,,
9466790,31920292,2018-07-16 12:29:00,Fabric Care,Detergent,minimarket,18500.0,0,1,7,July,2018,Monday,0,No,2018Q3,2018-07,16,12,18500.0,July / 12:29
9717406,32643087,2018-07-27 13:10:00,Fabric Care,Detergent,supermarket,32400.0,0,1,7,July,2018,Friday,4,No,2018Q3,2018-07,27,13,32400.0,July / 13:10
9371251,31628131,2018-07-11 20:29:00,Fabric Care,Detergent,minimarket,10700.0,0,1,7,July,2018,Wednesday,2,No,2018Q3,2018-07,11,20,10700.0,July / 20:29
9638365,32402239,2018-07-23 20:20:00,Fabric Care,Detergent,supermarket,5700.0,0,1,7,July,2018,Monday,0,No,2018Q3,2018-07,23,20,5700.0,July / 20:20
9312561,31586965,2018-07-04 21:10:00,Fabric Care,Detergent,minimarket,16000.0,0,1,7,July,2018,Wednesday,2,No,2018Q3,2018-07,4,21,16000.0,July / 21:10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5724525,17492369,2017-12-12 13:29:00,Fabric Care,Detergent,minimarket,21900.0,2000,1,12,December,2017,Tuesday,1,No,2017Q4,2017-12,12,13,19900.0,December / 13:29
5602515,17012060,2017-12-01 10:28:00,Fabric Care,Detergent,supermarket,5900.0,0,1,12,December,2017,Friday,4,No,2017Q4,2017-12,1,10,5900.0,December / 10:28
5880066,18161498,2017-12-24 16:49:00,Fabric Care,Detergent,minimarket,16200.0,0,1,12,December,2017,Sunday,6,Yes,2017Q4,2017-12,24,16,16200.0,December / 16:49


In [129]:
pd.crosstab(index=df['format'], columns=df['sub_category'], normalize='index')
## Normalize = index ==> Values di Index yg sama jika dijumlah = 1 => 100%

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,0.486129,0.185999,0.327872
minimarket,0.520159,0.151443,0.328398
supermarket,0.456169,0.197367,0.346464


In [130]:
pd.crosstab(index=df['format'], columns=df['sub_category'], normalize='index') * 100

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,48.612921,18.599888,32.787190
minimarket,52.015896,15.144328,32.839775
supermarket,45.616867,19.736709,34.646424


In [131]:
(pd.crosstab(index=df['format'], columns=df['sub_category'], normalize='index') * 100).round(2)

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,48.61,18.60,32.79
minimarket,52.02,15.14,32.84
supermarket,45.62,19.74,34.65


In [132]:
## hypermarket
48.61 + 18.60 + 32.79

100.0

In [133]:
df[df['format'] == 'hypermarket']

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month,bulan,year,day,hari,WeekEnd,quarter,yearMonth,tgl,hour,total_payment,custom
receipt_id,,,,,,,,,,,,,,,,,,,,
9763483,32856560,2018-07-31 05:51:00,Rice,Rice,hypermarket,77750.0,0,1,7,July,2018,Tuesday,1,No,2018Q3,2018-07,31,5,77750.0,July / 05:51
9455997,31889713,2018-07-15 20:01:00,Rice,Rice,hypermarket,64000.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,20,64000.0,July / 20:01
9281162,31442329,2018-07-07 21:15:00,Rice,Rice,hypermarket,64000.0,0,2,7,July,2018,Saturday,5,Yes,2018Q3,2018-07,7,21,128000.0,July / 21:15
9174690,31034105,2018-07-02 20:01:00,Rice,Rice,hypermarket,64000.0,0,2,7,July,2018,Monday,0,No,2018Q3,2018-07,2,20,128000.0,July / 20:01
9445585,31952944,2018-07-15 12:27:00,Rice,Rice,hypermarket,60900.0,0,2,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,12,121800.0,July / 12:27
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5872479,17910741,2017-12-24 13:14:00,Sugar/Flavored Syrup,Sugar,hypermarket,11500.0,6000,6,12,December,2017,Sunday,6,Yes,2017Q4,2017-12,24,13,63000.0,December / 13:14
5797765,17621628,2017-12-18 20:14:00,Sugar/Flavored Syrup,Sugar,hypermarket,12500.0,0,2,12,December,2017,Monday,0,No,2017Q4,2017-12,18,20,25000.0,December / 20:14
5678339,17317824,2017-12-08 17:11:00,Sugar/Flavored Syrup,Sugar,hypermarket,11990.0,260,1,12,December,2017,Friday,4,No,2017Q4,2017-12,8,17,11730.0,December / 17:11


In [134]:
(pd.crosstab(index=df['format'], columns=df['sub_category'], normalize='index') * 100).round(2).loc[['hypermarket']]

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,48.61,18.6,32.79


In [145]:
pd.crosstab(index=df['format'], columns=df['sub_category'], margins=True, margins_name = 'Total Transaksi')

sub_category,Detergent,Rice,Sugar,Total Transaksi
format,,,,
hypermarket,2611,999,1761,5371
minimarket,24345,7088,15370,46803
supermarket,9044,3913,6869,19826
Total Transaksi,36000,12000,24000,72000


In [148]:
pd.crosstab(index=df['format'], columns=df['sub_category'], margins=True).drop('All', axis=1)

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,2611,999,1761
minimarket,24345,7088,15370
supermarket,9044,3913,6869
All,36000,12000,24000


In [137]:
pd.crosstab(index=df['format'], columns='Jumlah')

col_0,Jumlah
format,
hypermarket,5371
minimarket,46803
supermarket,19826


In [138]:
pd.crosstab(index=df['sub_category'], columns='Jumlah')

col_0,Jumlah
sub_category,
Detergent,36000
Rice,12000
Sugar,24000


In [142]:
(pd.crosstab(index=df['format'], columns=df['sub_category'], margins=True, normalize='index') * 100).round(2)

sub_category,Detergent,Rice,Sugar
format,,,
hypermarket,48.61,18.60,32.79
minimarket,52.02,15.14,32.84
supermarket,45.62,19.74,34.65
All,50.00,16.67,33.33


In [143]:
(pd.crosstab(index=df['format'], columns=df['sub_category'], margins=True, normalize='columns') * 100).round(2)

sub_category,Detergent,Rice,Sugar,All
format,,,,
hypermarket,7.25,8.33,7.34,7.46
minimarket,67.62,59.07,64.04,65.00
supermarket,25.12,32.61,28.62,27.54


In [149]:
df.head()

,receipts_item_id,purchase_time,category,sub_category,format,unit_price,discount,quantity,month,bulan,year,day,hari,WeekEnd,quarter,yearMonth,tgl,hour,total_payment,custom
receipt_id,,,,,,,,,,,,,,,,,,,,
9622257,32369294,2018-07-22 21:19:00,Rice,Rice,supermarket,128000.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,22,21,128000.0,July / 21:19
9446359,31885876,2018-07-15 16:17:00,Rice,Rice,minimarket,102750.0,0,1,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,16,102750.0,July / 16:17
9470290,31930241,2018-07-15 12:12:00,Rice,Rice,supermarket,64000.0,0,3,7,July,2018,Sunday,6,Yes,2018Q3,2018-07,15,12,192000.0,July / 12:12
9643416,32418582,2018-07-24 08:27:00,Rice,Rice,minimarket,65000.0,0,1,7,July,2018,Tuesday,1,No,2018Q3,2018-07,24,8,65000.0,July / 08:27
9692093,32561236,2018-07-26 11:28:00,Rice,Rice,supermarket,124500.0,0,1,7,July,2018,Thursday,3,No,2018Q3,2018-07,26,11,124500.0,July / 11:28
